File: Source/Private Messages/MessageComposeViewController.m
//  MessageComposeViewController.m
//
//  Copyright 2013 Awful Contributors. CC BY-NC-SA 3.0 US https://github.com/Awful/Awful.app

#import "MessageComposeViewController.h"
#import "AwfulAppDelegate.h"
#import "AwfulForumsClient.h"
#import "AwfulFrameworkCategories.h"
#import "AwfulNewPrivateMessageFieldView.h"
#import "AwfulThreadTag.h"
#import "AwfulThreadTagLoader.h"
#import "AwfulThreadTagPickerController.h"
#import "Awful-Swift.h"

@interface MessageComposeViewController () <AwfulThreadTagPickerControllerDelegate, UIViewControllerRestoration>

@property (strong, nonatomic) AwfulNewPrivateMessageFieldView *fieldView;

@property (strong, nonatomic) AwfulThreadTag *threadTag;
@property (strong, nonatomic) AwfulThreadTagPickerController *threadTagPicker;

@property (assign, nonatomic) BOOL updatingThreadTags;
@property (copy, nonatomic) NSArray *availableThreadTags;

@end

@implementation MessageComposeViewController

- (id)initWithRecipient:(AwfulUser *)recipient
{
    if ((self = [self initWithNibName:nil bundle:nil])) {
        _recipient = recipient;
    }
    return self;
}

- (id)initWithRegardingMessage:(AwfulPrivateMessage *)regardingMessage initialContents:(NSString *)initialContents
{
    if ((self = [self initWithNibName:nil bundle:nil])) {
        _regardingMessage = regardingMessage;
        _initialContents = [initialContents copy];
    }
    return self;
}

- (id)initWithForwardingMessage:(AwfulPrivateMessage *)forwardingMessage initialContents:(NSString *)initialContents
{
    if ((self = [self initWithNibName:nil bundle:nil])) {
        _forwardingMessage = forwardingMessage;
        _initialContents = [initialContents copy];
    }
    return self;
}

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    if ((self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil])) {
        self.title = @"Private Message";
        self.submitButtonItem.title = @"Send";
        self.restorationClass = self.class;
    }
    return self;
}

- (void)setThreadTag:(AwfulThreadTag *)threadTag
{
    _threadTag = threadTag;
    [self updateThreadTagButtonImage];
}

- (void)loadView
{
    [super loadView];
    self.customView = self.fieldView;
}

- (void)themeDidChange
{
    [super themeDidChange];
    self.fieldView.toField.textField.textColor = self.textView.textColor;
    self.fieldView.toField.textField.keyboardAppearance = self.textView.keyboardAppearance;
    self.fieldView.subjectField.textField.textColor = self.textView.textColor;
    self.fieldView.subjectField.textField.keyboardAppearance = self.textView.keyboardAppearance;
    
    NSDictionary *styleAttrs = @{NSForegroundColorAttributeName: self.theme[@"placeholderTextColor"]};
    NSAttributedString *toString = [[NSAttributedString alloc] initWithString:@"To" attributes:styleAttrs];
    self.fieldView.toField.textField.attributedPlaceholder = toString;
    NSAttributedString *subjectString = [[NSAttributedString alloc] initWithString:@"Subject" attributes:styleAttrs];
    self.fieldView.subjectField.textField.attributedPlaceholder = subjectString;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    [self updateThreadTagButtonImage];
    if (self.recipient) {
        if (self.fieldView.toField.textField.text.length == 0) {
            self.fieldView.toField.textField.text = self.recipient.username;
        }
    } else if (self.regardingMessage) {
        if (self.fieldView.toField.textField.text.length == 0) {
            self.fieldView.toField.textField.text = self.regardingMessage.from.username;
        }
        if (self.fieldView.subjectField.textField.text.length == 0) {
            NSString *subject = self.regardingMessage.subject;
            if (![subject hasPrefix:@"Re: "]) {
                subject = [NSString stringWithFormat:@"Re: %@", subject];
            }
            self.fieldView.subjectField.textField.text = subject;
        }
        if (self.textView.text.length == 0) {
            self.textView.text = self.initialContents;
        }
    } else if (self.forwardingMessage) {
        if (self.fieldView.subjectField.textField.text.length == 0) {
            self.fieldView.subjectField.textField.text = [NSString stringWithFormat:@"Fw: %@", self.forwardingMessage.subject];
        }
        if (self.textView.text.length == 0) {
            self.textView.text = self.initialContents;
        }
    }
    
    [self updateAvailableThreadTagsIfNecessary];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [self updateAvailableThreadTagsIfNecessary];
}

- (void)updateAvailableThreadTagsIfNecessary
{
    if (self.availableThreadTags.count > 0 || self.updatingThreadTags) return;
    self.updatingThreadTags = YES;
    __weak __typeof__(self) weakSelf = self;
    [[AwfulForumsClient client] listAvailablePrivateMessageThreadTagsAndThen:^(NSError *error, NSArray *threadTags) {
        __typeof__(self) self = weakSelf;
        self.availableThreadTags = threadTags;
        self.updatingThreadTags = NO;
    }];
}

- (AwfulNewPrivateMessageFieldView *)fieldView
{
    if (_fieldView) return _fieldView;
    _fieldView = [[AwfulNewPrivateMessageFieldView alloc] initWithFrame:CGRectMake(0, 0, 0, 88)];
    _fieldView.toField.label.textColor = [UIColor grayColor];
    _fieldView.subjectField.label.textColor = [UIColor grayColor];
    [_fieldView.threadTagButton addTarget:self
                                   action:@selector(didTapThreadTagButton:)
                         forControlEvents:UIControlEventTouchUpInside];
    [_fieldView.toField.textField addTarget:self
                                     action:@selector(toFieldDidChange)
                           forControlEvents:UIControlEventEditingChanged];
    [_fieldView.subjectField.textField addTarget:self
                                          action:@selector(subjectFieldDidChange)
                                forControlEvents:UIControlEventEditingChanged];
    return _fieldView;
}

- (void)updateThreadTagButtonImage
{
    UIImage *image;
    if (self.threadTag.imageName) {
        image = [AwfulThreadTagLoader imageNamed:self.threadTag.imageName];
    } else {
        image = [AwfulThreadTagLoader unsetThreadTagImage];
    }
    [self.fieldView.threadTagButton setImage:image forState:UIControlStateNormal];
}

- (void)didTapThreadTagButton:(AwfulThreadTagButton *)button
{
    // TODO better handle waiting for available thread tags to download.
    if (self.availableThreadTags.count == 0) return;
    
    NSString *selectedImageName = self.threadTag.imageName ?: AwfulThreadTagLoaderEmptyPrivateMessageImageName;
    [self.threadTagPicker selectImageName:selectedImageName];
    [self.threadTagPicker presentFromView:button];
    
    // HACK: Calling -endEditing: once doesn't work if the To or Subject fields are selected. But twice works. I assume this is some weirdness from adding text fields as subviews to a text view.
    [self.view endEditing:YES];
    [self.view endEditing:YES];
}

- (AwfulThreadTagPickerController *)threadTagPicker
{
    if (_threadTagPicker) return _threadTagPicker;
    if (self.availableThreadTags.count == 0) return nil;
    
    NSMutableArray *imageNames = [NSMutableArray arrayWithObject:AwfulThreadTagLoaderEmptyPrivateMessageImageName];
    [imageNames addObjectsFromArray:[self.availableThreadTags valueForKey:@"imageName"]];
    _threadTagPicker = [[AwfulThreadTagPickerController alloc] initWithImageNames:imageNames secondaryImageNames:nil];
    _threadTagPicker.delegate = self;
    _threadTagPicker.navigationItem.leftBarButtonItem = _threadTagPicker.cancelButtonItem;
    return _threadTagPicker;
}

- (void)toFieldDidChange
{
    [self updateSubmitButtonItem];
}

- (void)subjectFieldDidChange
{
    [self updateSubmitButtonItem];
}

- (BOOL)canSubmitComposition
{
    return ([super canSubmitComposition] &&
            self.fieldView.toField.textField.text.length > 0 &&
            self.fieldView.subjectField.textField.text.length > 0);
}

- (NSString *)submissionInProgressTitle
{
    return @"Sending…";
}

- (void)submitComposition:(NSString *)composition completionHandler:(void (^)(BOOL))completionHandler
{
    __weak __typeof__(self) weakSelf = self;
    [[AwfulForumsClient client] sendPrivateMessageTo:self.fieldView.toField.textField.text
                                         withSubject:self.fieldView.subjectField.textField.text
                                           threadTag:self.threadTag
                                              BBcode:composition
                                    asReplyToMessage:self.regardingMessage
                                forwardedFromMessage:self.forwardingMessage
                                             andThen:^(NSError *error)
    {
        __typeof__(self) self = weakSelf;
        if (error) {
            completionHandler(NO);
            [self presentViewController:[UIAlertController alertWithNetworkError:error] animated:YES completion:nil];
        } else {
            completionHandler(YES);
        }
    }];
}

#pragma mark - AwfulThreadTagPickerControllerDelegate

- (void)threadTagPicker:(AwfulThreadTagPickerController *)picker didSelectImageName:(NSString *)imageName
{
    if ([imageName isEqualToString:AwfulThreadTagLoaderEmptyPrivateMessageImageName]) {
        self.threadTag = nil;
    } else {
        [self.availableThreadTags enumerateObjectsUsingBlock:^(AwfulThreadTag *threadTag, NSUInteger i, BOOL *stop) {
            if ([threadTag.imageName isEqualToString:imageName]) {
                self.threadTag = threadTag;
                *stop = YES;
            }
        }];
    }
    [picker dismiss];
    [self focusInitialFirstResponder];
}

#pragma mark - UIViewControllerRestoration

+ (UIViewController *)viewControllerWithRestorationIdentifierPath:(NSArray *)identifierComponents coder:(NSCoder *)coder
{
    NSString *recipientUserID = [coder decodeObjectForKey:RecipientUserIDKey];
    NSString *regardingMessageID = [coder decodeObjectForKey:RegardingMessageIDKey];
    NSString *forwardingMessageID = [coder decodeObjectForKey:ForwardingMessageIDKey];
    NSString *initialContents = [coder decodeObjectForKey:InitialContentsKey];
    NSManagedObjectContext *managedObjectContext = [AwfulAppDelegate instance].dataStack.managedObjectContext;
    MessageComposeViewController *newPrivateMessageViewController;
    if (recipientUserID) {
        AwfulUser *recipient = [AwfulUser firstOrNewUserWithUserID:recipientUserID
                                                          username:nil
                                            inManagedObjectContext:managedObjectContext];
        newPrivateMessageViewController = [[MessageComposeViewController alloc] initWithRecipient:recipient];
    } else if (regardingMessageID) {
        AwfulPrivateMessage *regardingMessage = [AwfulPrivateMessage fetchArbitraryInManagedObjectContext:managedObjectContext
                                                                                  matchingPredicateFormat:@"messageID = %@", regardingMessageID];
        newPrivateMessageViewController = [[MessageComposeViewController alloc] initWithRegardingMessage:regardingMessage
                                                                                                 initialContents:initialContents];
    } else if (forwardingMessageID) {
        AwfulPrivateMessage *forwardingMessage = [AwfulPrivateMessage fetchArbitraryInManagedObjectContext:managedObjectContext
                                                                                   matchingPredicateFormat:@"messageID = %@", forwardingMessageID];
        newPrivateMessageViewController = [[MessageComposeViewController alloc] initWithForwardingMessage:forwardingMessage
                                                                                                  initialContents:initialContents];
    } else {
        newPrivateMessageViewController = [[MessageComposeViewController alloc] initWithRecipient:nil];
    }
    newPrivateMessageViewController.restorationIdentifier = identifierComponents.lastObject;
    NSError *error;
    if (![managedObjectContext save:&error]) {
        NSLog(@"%s error saving managed object context: %@", __PRETTY_FUNCTION__, error);
    }
    return newPrivateMessageViewController;
}

- (void)encodeRestorableStateWithCoder:(NSCoder *)coder
{
    [super encodeRestorableStateWithCoder:coder];
    [coder encodeObject:self.recipient.userID forKey:RecipientUserIDKey];
    [coder encodeObject:self.regardingMessage.messageID forKey:RegardingMessageIDKey];
    [coder encodeObject:self.forwardingMessage forKey:ForwardingMessageIDKey];
    [coder encodeObject:self.initialContents forKey:InitialContentsKey];
    [coder encodeObject:self.threadTag.imageName forKey:ThreadTagImageNameKey];
}

- (void)decodeRestorableStateWithCoder:(NSCoder *)coder
{
    self.fieldView.toField.textField.text = [coder decodeObjectForKey:ToKey];
    self.fieldView.subjectField.textField.text = [coder decodeObjectForKey:SubjectKey];
    NSString *threadTagImageName = [coder decodeObjectForKey:ThreadTagImageNameKey];
    if (threadTagImageName) {
        NSManagedObjectContext *managedObjectContext = (self.recipient.managedObjectContext ?:
                                                        self.regardingMessage.managedObjectContext ?:
                                                        self.forwardingMessage.managedObjectContext);
        self.threadTag = [AwfulThreadTag firstOrNewThreadTagWithThreadTagID:nil
                                                                  imageName:threadTagImageName
                                                     inManagedObjectContext:managedObjectContext];
    }
    
    [super decodeRestorableStateWithCoder:coder];
}

static NSString * const RecipientUserIDKey = @"AwfulRecipientUserID";
static NSString * const RegardingMessageIDKey = @"AwfulRegardingMessageID";
static NSString * const ForwardingMessageIDKey = @"AwfulForwardingMessageID";
static NSString * const InitialContentsKey = @"AwfulInitialContents";
static NSString * const ToKey = @"AwfulTo";
static NSString * const SubjectKey = @"AwfulSubject";
static NSString * const ThreadTagImageNameKey = @"AwfulThreadTagImageName";

@end

File: App/View Controllers/Messages/MessageComposeViewController.swift
//  MessageComposeViewController.swift
//
//  Copyright 2016 Awful Contributors. CC BY-NC-SA 3.0 US https://github.com/Awful/Awful.app

import AwfulCore
import Nuke
import os
import UIKit

private let logger = Logger(subsystem: Bundle.main.bundleIdentifier!, category: "MessageComposeViewController")

/// For writing private messages.
final class MessageComposeViewController: ComposeTextViewController {
    fileprivate let recipient: User?
    fileprivate let regardingMessage: PrivateMessage?
    fileprivate let forwardingMessage: PrivateMessage?
    fileprivate let initialContents: String?
    private let draft: PrivateMessageDraft
    private var autoSaveWorkItem: DispatchWorkItem?
    fileprivate var threadTag: ThreadTag? {
        didSet { updateThreadTagButtonImage() }
    }
    fileprivate var availableThreadTags: [ThreadTag]?
    fileprivate var updatingThreadTags = false
    fileprivate var threadTagPicker: ThreadTagPickerViewController?
    
    fileprivate lazy var fieldView: NewPrivateMessageFieldView = {
        let fieldView = NewPrivateMessageFieldView(frame: CGRect(x: 0, y: 0, width: 0, height: 88))
        fieldView.toField.label.textColor = .gray
        fieldView.subjectField.label.textColor = .gray
        fieldView.threadTagButton.addTarget(self, action: #selector(didTapThreadTagButton), for: .touchUpInside)
        fieldView.toField.textField.addTarget(self, action: #selector(toFieldDidChange), for: .editingChanged)
        fieldView.subjectField.textField.addTarget(self, action: #selector(subjectFieldDidChange), for: .editingChanged)
        return fieldView
    }()
    
    init() {
        recipient = nil
        regardingMessage = nil
        forwardingMessage = nil
        initialContents = nil
        self.draft = Self.loadDraft(kind: .new)
        super.init(nibName: nil, bundle: nil)
        commonInit()
    }

    init(recipient: User) {
        self.recipient = recipient
        regardingMessage = nil
        forwardingMessage = nil
        initialContents = nil
        self.draft = Self.loadDraft(kind: .to(recipient))
        super.init(nibName: nil, bundle: nil)
        commonInit()
    }

    init(regardingMessage: PrivateMessage, initialContents: String?) {
        self.regardingMessage = regardingMessage
        recipient = nil
        forwardingMessage = nil
        self.initialContents = initialContents
        self.draft = Self.loadDraft(kind: .replying(to: regardingMessage))
        super.init(nibName: nil, bundle: nil)
        commonInit()
    }

    init(forwardingMessage: PrivateMessage, initialContents: String?) {
        self.forwardingMessage = forwardingMessage
        recipient = nil
        regardingMessage = nil
        self.initialContents = initialContents
        self.draft = Self.loadDraft(kind: .forwarding(forwardingMessage))
        super.init(nibName: nil, bundle: nil)
        commonInit()
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    deinit {
        autoSaveWorkItem?.cancel()
    }

    /// Returns a fresh draft if no on-disk draft exists for the given kind, otherwise the loaded
    /// draft. Resolving the on-disk path doesn't depend on the actual `PrivateMessage` instance,
    /// just its message ID, so this works at init time before any properties are set.
    private static func loadDraft(kind: PrivateMessageDraft.Kind) -> PrivateMessageDraft {
        let placeholder = PrivateMessageDraft(kind: kind)
        if let saved = DraftStore.sharedStore().loadDraft(placeholder.storePath) as? PrivateMessageDraft {
            return saved
        }
        return placeholder
    }

    fileprivate func commonInit() {
        title = "Private Message"
        submitButtonItem.title = "Send"
    }
    
    private var threadTagTask: ImageTask?
    
    fileprivate func updateThreadTagButtonImage() {
        threadTagTask?.cancel()
        fieldView.threadTagButton.setImage(ThreadTagLoader.Placeholder.thread(tintColor: nil).image, for: .normal)
        
        threadTagTask = ThreadTagLoader.shared.loadImage(named: threadTag?.imageName) { [fieldView] result in
            switch result {
            case .success(let response):
                fieldView.threadTagButton.setImage(response.image, for: .normal)
            case .failure:
                break
            }
        }
    }
    
    fileprivate func updateAvailableThreadTagsIfNecessary() {
        guard availableThreadTags?.isEmpty ?? true else { return }
        guard !updatingThreadTags else { return }
        Task {
            do {
                let threadTags = try await ForumsClient.shared.listAvailablePrivateMessageThreadTags()
                availableThreadTags = threadTags

                let picker = ThreadTagPickerViewController(
                    firstTag: .privateMessage,
                    imageNames: threadTags.compactMap { $0.imageName },
                    secondaryImageNames: [])
                picker.delegate = self
                picker.navigationItem.leftBarButtonItem = picker.cancelButtonItem
                threadTagPicker = picker
            } catch {
                logger.error("Could not list available private message thread tags: \(error)")
            }

            updatingThreadTags = false
        }
    }
    
    override var canSubmitComposition: Bool {
        return super.canSubmitComposition &&
        fieldView.toField.textField.text?.isEmpty == false &&
        fieldView.subjectField.textField.text?.isEmpty == false
    }
    
    override var submissionInProgressTitle: String {
        return "Sending…"
    }
    
    override func submit(_ composition: String, completion: @escaping (Bool) -> Void) {
        guard let to = fieldView.toField.textField.text,
              !to.isEmpty,
              let subject = fieldView.subjectField.textField.text,
              !subject.isEmpty
        else { return }
        Task {
            do {
                let relevant: ForumsClient.RelevantMessage
                if let regardingMessage {
                    relevant = .replyingTo(regardingMessage)
                } else if let forwardingMessage {
                    relevant = .forwarding(forwardingMessage)
                } else {
                    relevant = .none
                }
                try await ForumsClient.shared.sendPrivateMessage(to: to, subject: subject, threadTag: threadTag, bbcode: composition, about: relevant)
                deleteDraft()
                completion(true)
            } catch {
                completion(false)
                present(UIAlertController(networkError: error), animated: true)
            }
        }
    }
    
    @objc fileprivate func didTapThreadTagButton(_ sender: ThreadTagButton) {
        guard let picker = threadTagPicker else { return }
        
        picker.selectImageName(threadTag?.imageName)
        picker.present(from: self, sourceView: sender)
        
        // HACK: Calling -endEditing: once doesn't work if the To or Subject fields are selected. But twice works. I assume this is some weirdness from adding text fields as subviews to a text view.
        view.endEditing(true)
        view.endEditing(true)
    }
    
    @objc fileprivate func toFieldDidChange() {
        updateSubmitButtonItem()
        scheduleDraftAutoSave()
    }

    @objc fileprivate func subjectFieldDidChange() {
        updateSubmitButtonItem()
        scheduleDraftAutoSave()
    }

    override func bodyTextDidChange() {
        scheduleDraftAutoSave()
    }

    private func scheduleDraftAutoSave() {
        autoSaveWorkItem?.cancel()
        let work = DispatchWorkItem { [weak self] in self?.saveDraftNow() }
        autoSaveWorkItem = work
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.5, execute: work)
    }

    /// Synchronously runs any pending auto-save. Called on dismissal so the last keystrokes
    /// aren't lost to the 0.5 s debounce.
    private func flushDraftAutoSave() {
        guard autoSaveWorkItem != nil else { return }
        autoSaveWorkItem?.cancel()
        autoSaveWorkItem = nil
        saveDraftNow()
    }

    private func saveDraftNow() {
        draft.to = fieldView.toField.textField.text ?? ""
        draft.subject = fieldView.subjectField.textField.text ?? ""
        draft.threadTag = threadTag
        draft.text = textView.attributedText
        if draft.to.isEmpty
            && draft.subject.isEmpty
            && draft.threadTag == nil
            && (draft.text?.length ?? 0) == 0
        {
            DraftStore.sharedStore().deleteDraft(draft)
        } else {
            DraftStore.sharedStore().saveDraft(draft)
        }
    }

    private func deleteDraft() {
        autoSaveWorkItem?.cancel()
        autoSaveWorkItem = nil
        DraftStore.sharedStore().deleteDraft(draft)
    }
    
    // MARK: View lifecycle
    
    override func loadView() {
        super.loadView()
        customView = fieldView
    }
    
    override func themeDidChange() {
        super.themeDidChange()
        
        fieldView.toField.textField.textColor = textView.textColor
        fieldView.toField.textField.keyboardAppearance = textView.keyboardAppearance
        fieldView.subjectField.textField.textColor = textView.textColor
        fieldView.subjectField.textField.keyboardAppearance = textView.keyboardAppearance
        
        let attributes = [NSAttributedString.Key.foregroundColor: theme[uicolor: "placeholderTextColor"] ?? .gray]
        fieldView.toField.textField.attributedPlaceholder = NSAttributedString(string: "To", attributes: attributes)
        fieldView.subjectField.textField.attributedPlaceholder = NSAttributedString(string: "Subject", attributes: attributes)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()

        threadTag = draft.threadTag
        updateThreadTagButtonImage()

        // Saved-draft contents take precedence over the defaults derived from the recipient or
        // the message being replied to/forwarded — the draft already incorporates any user edits.
        if !draft.to.isEmpty {
            fieldView.toField.textField.text = draft.to
        } else if let recipient = recipient {
            fieldView.toField.textField.text = recipient.username
        } else if let regardingMessage = regardingMessage {
            fieldView.toField.textField.text = regardingMessage.from?.username
        }

        if !draft.subject.isEmpty {
            fieldView.subjectField.textField.text = draft.subject
        } else if let regardingMessage = regardingMessage {
            var subject = regardingMessage.subject ?? ""
            if !subject.hasPrefix("Re: ") {
                subject = "Re: \(subject)"
            }
            fieldView.subjectField.textField.text = subject
        } else if let forwardingMessage = forwardingMessage {
            fieldView.subjectField.textField.text = "Fw: \(forwardingMessage.subject ?? "")"
        }

        if let savedText = draft.text {
            textView.attributedText = savedText
            updateSubmitButtonItem()
        } else if let initialContents = initialContents, textView.text.isEmpty {
            textView.text = initialContents
        }

        updateAvailableThreadTagsIfNecessary()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)

        updateAvailableThreadTagsIfNecessary()
    }

    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)

        // Flush the debounced draft save on the way out so nothing the user typed in the last
        // 0.5 s is dropped if they dismiss immediately after typing.
        if isMovingFromParent || isBeingDismissed {
            flushDraftAutoSave()
        }
    }

}

extension MessageComposeViewController: ThreadTagPickerViewControllerDelegate {
    
    func didSelectImageName(_ imageName: String?, in picker: ThreadTagPickerViewController) {
        if
            let imageName = imageName,
            let availableThreadTags = availableThreadTags,
            let i = availableThreadTags.firstIndex(where: { $0.imageName == imageName })
        {
            threadTag = availableThreadTags[i]
        } else {
            threadTag = nil
        }
        scheduleDraftAutoSave()

        picker.dismiss()

        focusInitialFirstResponder()
    }
    
    func didSelectSecondaryImageName(_ secondaryImageName: String, in picker: ThreadTagPickerViewController) {
        // nop
    }
    
    func didDismissPicker(_ picker: ThreadTagPickerViewController) {
        // nop
    }
}

